Sources/FirebladeECS/FSM.swift

```diff
@@ -0,0 +1,422 @@
+    /// - Returns: This ComponentMapping, so more modifications can be applied
+    @discardableResult
+    public func withType(_ type: ComponentInitializable.Type) -> Self {
+        setProvider(ComponentTypeProvider(type: type))
+        return self
+    }
+
+    /// Creates a mapping for the component type to a single instance of the provided type.
+    /// The instance is not created until it is first requested. The type should be the same
+    /// as or extend the type for this mapping. A ComponentSingletonProvider is used for
+    /// the mapping.
+    /// - Parameter type: The type of the single instance to be created. If omitted, the type of the
+    /// mapping is used.
+    /// - Returns: This ComponentMapping, so more modifications can be applied
+    @discardableResult
+    public func withSingleton(_ type: ComponentInitializable.Type?) -> Self {
+        setProvider(ComponentSingletonProvider(type: type ?? componentType))
+        return self
+    }
+
+    /// Creates a mapping for the component type to a method call. A
+    /// DynamicComponentProvider is used for the mapping.
+    /// - Parameter closure: The Closure instance to return the component instance
+    /// - Returns: This ComponentMapping, so more modifications can be applied
+    @discardableResult
+    public func withMethod<C: Component>(_ closure: DynamicComponentProvider<C>.Closure) -> Self {
+        setProvider(DynamicComponentProvider(closure: closure))
+        return self
+    }
+
+    /// Creates a mapping for the component type to any ComponentProvider.
+    /// - Parameter provider: The component provider to use.
+    /// - Returns: This ComponentMapping, so more modifications can be applied.
+    @discardableResult
+    public func withProvider(_ provider: ComponentProvider) -> Self {
+        setProvider(provider)
+        return self
+    }
+
+    /// Maps through to the add method of the EntityState that this mapping belongs to
+    /// so that a fluent interface can be used when configuring entity states.
+    /// - Parameter type: The type of component to add a mapping to the state for
+    /// - Returns: The new ComponentMapping for that type
+    @discardableResult
+    public func add(_ type: ComponentInitializable.Type) -> StateComponentMapping {
+        creatingState.add(type)
+    }
+
+    private func setProvider(_ provider: ComponentProvider) {
+        self.provider = provider
+        creatingState.providers[componentType.identifier] = provider
+    }
+}
+
+// MARK: -
+
+/// This is a state machine for an entity. The state machine manages a set of states,
+/// each of which has a set of component providers. When the state machine changes the state, it removes
+/// components associated with the previous state and adds components associated with the new state.
+/// - Parameter StateName: Generic hashable state name type
+public class EntityStateMachine<StateName: Hashable> {
```

Comment:
Maybe we should conform the `StateName` to the [Identifiable](https://developer.apple.com/documentation/swift/identifiable) protocol since it would allow collection diffing and give a clearer understanding of the use of this type. Furthermore it's implementations should guarantee the following:

> Identifiable leaves the duration and scope of the identity unspecified. Identities could be any of the following:
> - Guaranteed always unique (e.g. UUIDs).
> - Persistently unique per environment (e.g. database record keys).
> - Unique for the lifetime of a process (e.g. global incrementing integers).
> - Unique for the lifetime of an object (e.g. object identifiers).
> - Unique within the current collection (e.g. collection index).

All these criteria apply to a state name as far as I understand.
Furthermore I would recommend to rename `StateName` to `StateIdentifier` here 🤷  
What do you think?

Reply:
StateName (from now on StateIdentifier) may conform to Identifiable. My idea was to make StateIdentifier generic to allow more flexibility for choosing type that represents name (using enum for example). The only problem I face with Identifiable right now is that it is [deployment target specific](https://forums.swift.org/t/identifiable-restricted-to-ios-13-for-spm-packages/30105), so I get _"'Identifiable' is only available in macOS 10.15 or newer"_ error in Xcode.  Any idea how to remedy this?

Reply:
You are right, I forgot about the (unnecessary) platform constraints `@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)` on the [Identifiable protocol implementation](https://github.com/apple/swift/blob/7df8b7061de7b644305d352ffcf55cd4991db864/stdlib/public/core/Identifiable.swift#L41) 🙈 

I tried several ways to work around it, but failed. 
Providing a custom implementation for os versions less than a specific version is not possible with Swift afaik, sadly.
This is due to the fact that `#available` is only available inside function bodies and not on a root level. Furthermore `#if os(...)` does not take an os version comparator.

On the other hand we do not want to constrain our target platforms just for one (simple) protocol conformance.
As of now FirebladeECS works on all available os version that support Swift 5.1 and higher, which I find more important than the `Identifiable` protocol conformance.
Therefore I would recommend we abandon the `Identifiable` conformance and keep `Hashable.

Do you agree?

Reply:
Totally. Indeed there's nothing (in my opinion) runtime/compiler specific for this protocol to have such constraints. 

With this change addressed do you have any other suggestions that may improve this PR?

Reply:
No, I do not. 
I've you do not have anything else to add, I would like to merge your excellent work 👍